src/services/datasource.ts

```diff
@@ -25,26 +45,85 @@ class WrappedLokiDatasource extends RuntimeDataSource<DataQuery> {
       getDataSourceSrv()
         .get(getDataSource(request.scopedVars.__sceneObject.valueOf()))
         .then((ds) => {
-          // override the target datasource to Loki
-          request.targets = request.targets.map((target) => {
-            target.datasource = ds;
-            return target;
-          });
-
-          // query the datasource and return either observable or promise
-          const dsResponse = ds.query(request);
-          if (isObservable(dsResponse)) {
-            dsResponse.subscribe(subscriber);
-          } else {
-            dsResponse.then((response) => {
-              subscriber.next(response);
-              subscriber.complete();
-            });
+          if (!(ds instanceof DataSourceWithBackend)) {
+            throw new Error('Invalid datasource!');
+          }
+
+          const requestType = request.targets?.[0]?.resource;
+
+          switch (requestType) {
+            case 'volume': {
+              this.transformVolumeResponse(request, ds, subscriber);
+              break;
+            }
+            case 'patterns': {
+              console.warn('not yet implemented');
+              // this.transformPatternResponse(request, ds, subscriber);
+              break;
+            }
+            default: {
+              // override the target datasource to Loki
+              request.targets = request.targets.map((target) => {
+                target.datasource = ds;
+                return target;
+              });
+
+              // query the datasource and return either observable or promise
+              const dsResponse = ds.query(request);
+              dsResponse.subscribe(subscriber);
+              break;
+            }
           }
         });
     });
   }
 
+  private transformVolumeResponse(
+    request: DataQueryRequest<LokiQuery & SceneDataQueryResourceRequest>,
+    ds: DataSourceWithBackend<LokiQuery>,
+    subscriber: Subscriber<DataQueryResponse>
+  ) {
+    if (request.targets.length !== 1) {
+      throw new Error('Volume query can only have a single target!');
+    }
+
+    request.targets.forEach((target) => {
+      target.expr = target.expr.replace(VAR_SERVICE_EXPR_HACK, '');
+    });
+
+    const targetsInterpolated = ds.interpolateVariablesInQueries(request.targets, request.scopedVars);
+
+    const dsResponse = ds.getResource(
+      'index/volume',
+      {
+        expr: targetsInterpolated[0].expr.replace(VAR_SERVICE_EXPR_HACK, ''),
+        query: targetsInterpolated[0].expr.replace(VAR_SERVICE_EXPR_HACK, ''),
+        from: request.range.from.utc().toISOString(),
+        to: request.range.to.utc().toISOString(),
+        limit: 1000,
+      },
+      {
+        headers: {
```

Comment:
```suggestion
        requestId: request.requestId ?? 'volume',
        headers: {
```

This gets us automatic request cancellation. I noticed, when switching datasource, those inflight requests were not cancelled.

Reply:
Great catch, can confirm, thank you Sven!